v2/internal/controllers/generic_controller.go

```diff
@@ -289,7 +301,8 @@ func (gr *GenericReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		gr.ResourceResolver,
 		gr.PositiveConditions,
 		gr.Config,
-		gr.rand)
+		gr.rand,
```

Comment:
This constructor pulls a lot of things off gr to pass in - could it just take a reference to gr and use the fields directly? I guess that would couple them, but it kind of seems like they're coupled anyway.

Reply:
I agree, it does kinda feel like the object-envy code smell.

Reply:
We can't pass `GenericReconciler` here because that would introduce a cyclic dependency, where `GenericReconciler` depends on `AzureDeploymentReconciler` (like it does already), but `AzureDeploymentReconciler` depends on `GenericReconciler`.

The intent/goal (that has never been fully realized) is:
`GenericReconciler`: Implements `reconcile.Reconciler` (the controller-runtime interface). Performs some logging setup, gets the initial object and ensures it implements `genruntime.MetaObject` (our base interface all types must implement). Defers actual reconciliation to a separate interface `genruntime.Reconciler`.

`AzureDeploymentReconciler`: Generic reconciler that uses ARM PUT to reconcile resources. Implements `genruntime.Reconciler`. Is provided with logging and a `genruntime.MetaObject`.

There could be other implementers of `genruntime.Reconciler` that took a `genruntime.MetaObject` but reconciled them differently (for say, handcrafted resources that cannot use the ARM API such as SQL users). These would still be a `genruntime.MetaObject` and have the same logging, etc setups, but the actual reconcile logic would differ.

The way it probably _ought_ to work is that when we register resource types we also have an affiliated reconciler type that we infer somehow, and when we construct the `GenericReconciler` we would also include the associated `genruntime.Reconciler`. This way you have an outer reconciler that's the same for all resources, but an inner one that can differ. The outer ensures uniform logging setup, etc while the inner is free to do as it pleases.

This is part of what the comment above saying `TODO: We need some factory-lookup here` is getting at. Really `GenericReconciler` probably should not be new-ing up the `AzureDeploymentReconciler` because that pins it to only a single inner impl when we really want many possible.